File: CPE/Star/3/UVA-11151 Longest Palindrome.cpp
/*
    Longest Palindrome Subsequence
    1. 轉換為 LCS 問題
    2. 區間 DP (Bottom-Up) 枚舉長度，由小區間到大區間
*/
#include <bits/stdc++.h>
using namespace std;
const int N = 1024;
#define endl '\n'

string s, t;
int dp[N][N];

void solve1() {
    int n = s.size();
    string t = s;
    reverse(t.begin(), t.end());
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= n; ++j) {
            if (s[i - 1] == t[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
            else dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
        }
    }
    cout << dp[n][n] << endl;
}

void solve2() {
    int n = s.size();
    if (n == 0) {
        cout << 0 << endl;
        return;
    }
    for (int i = 0; i < n; ++i) dp[i][i] = 1;
    for (int ln = 2; ln <= n; ++ln) {
        for (int i = 0; i <= n - ln; ++i) {
            int j = i + ln - 1;
            if (s[i] == s[j]) dp[i][j] = dp[i + 1][j - 1] + 2;
            else dp[i][j] = max(dp[i + 1][j], dp[i][j - 1]);
        }
    }
    cout << dp[0][n - 1] << endl;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int T;
    cin >> T;
    cin.ignore();
    while (T--) {
        getline(cin, s);
        memset(dp, 0, sizeof(dp));
        solve1();
        // solve2();
    }
    return 0;
}

File: CPE/Star/3/UVA-11151 Longest Palindrome.py
"""
    Longest Palindrome Subsequence
    1. 轉換為 LCS 問題
    2. 區間 DP (Bottom-Up) 枚舉長度，由小區間到大區間
"""
T = int(input())

def solve1():
    s = input()
    t = s[::-1]
    n = len(s)
    f = [[0] * (n + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if s[i - 1] == t[j - 1]:
                f[i][j] = f[i - 1][j - 1] + 1
            else:
                f[i][j] = max(f[i - 1][j], f[i][j - 1])
    print(f[n][n])

def solve2():
    s = input()
    n = len(s)
    if n == 0: # 特判輸入為空行的情況
        print(0)
        return
    f = [[0] * n for _ in range(n)]
    for i in range(n): # 初始化區間長度為 1 的值
        f[i][i] = 1
    for ln in range(2, n + 1): # 枚舉區間長度
        for i in range(n - ln + 1): # 枚舉區間左端點
            j = i + ln - 1 # 區間右端點
            if s[i] == s[j]:
                f[i][j] = f[i + 1][j - 1] + 2
            else:
                f[i][j] = max(f[i + 1][j], f[i][j - 1])
    print(f[0][n - 1])

for _ in range(T):
    # solve1()
    solve2()